### PR TITLE
chore: reorg repo level utils, lint and typecheck repo files

### PR DESCRIPTION
**Pull request by JamesHenry**
Supersedes #8880

Explainer/walkthrough for this PR: https://www.loom.com/share/2c128b6caecb420abda03871f3043bbb

### CODE REVIEW

**Comment by abrahamguo** [comment]
Nice! Thanks a bunch for doing this and putting together the detailed walkthrough.

I admire people who are knowledgeable and interested in the parts I'm not, like repo setup with`nx` haha

File: tsconfig.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "compilerOptions": {
     "types": ["@types/node"],
     "noEmit": true,
-    "allowJs": true
+    "allowJs": true,
+    "allowImportingTsExtensions": true
```
Comment (abrahamguo) [inline tsconfig.json:6]: why is this needed?

File: .github/actions/breaking-pr-check/index.js [hunk 2]
```diff
@@ -1,12 +1,17 @@
+// @ts-check
+/* eslint-disable jsdoc/no-types */
```
Comment (JoshuaKGoldberg) [inline .github/actions/breaking-pr-check/index.js:2]: This makes me think we should change `eslint.config.mjs` to only apply the TypeScript rules to TypeScript files. That's fine as a followup IMO.

Comment (JoshuaKGoldberg) [inline .github/actions/breaking-pr-check/index.js:2]: This makes me think we should change `eslint.config.mjs` to only apply the TypeScript rules to TypeScript files. That's fine as a followup IMO.

File: packages/repo-tools/CHANGELOG.md
Comment (JoshuaKGoldberg) [inline packages/repo-tools/CHANGELOG.md:1]: 😂 yes, good.

File: tsconfig.json (hunk 1 above)
Comment (JoshuaKGoldberg) [reply tsconfig.json:6]: Without it, `yarn typecheck`'s `tsc` pass fails:

```plaintext
   ✖  nx run repo:typecheck
      > tsc
      
      tools/scripts/generate-configs.mts:18:8 - error TS5097: An import path can only end with a '.mts' extension when 'allowImportingTsExtensions' is enabled.
      
      18 } from './paths.mts';
                ~~~~~~~~~~~~~
      
      tools/scripts/generate-contributors.mts:10:27 - error TS5097: An import path can only end with a '.mts' extension when 'allowImportingTsExtensions' is enabled.
      
      10 import { REPO_ROOT } from './paths.mts';
                                   ~~~~~~~~~~~~~
      
      tools/scripts/generate-lib.mts:24:8 - error TS5097: An import path can only end with a '.mts' extension when 'allowImportingTsExtensions' is enabled.
      
      24 } from './paths.mts';
                ~~~~~~~~~~~~~
      
      tools/scripts/generate-sponsors.mts:6:34 - error TS5097: An import path can only end with a '.mts' extension when 'allowImportingTsExtensions' is enabled.
      
      6 import { PACKAGES_WEBSITE } from './paths.mts';
                                         ~~~~~~~~~~~~~
      
      
      Found 4 errors in 4 files.
      
      Errors  Files
           1  tools/scripts/generate-configs.mts:18
           1  tools/scripts/generate-contributors.mts:10
           1  tools/scripts/generate-lib.mts:24
           1  tools/scripts/generate-sponsors.mts:6
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
Looks great to me, thanks @JamesHenry! Excited to see one less CHANGELOG.md 😄

File: tsconfig.json (hunk 1 above)
Comment (JamesHenry) [reply tsconfig.json:6]: Yeah this was an option set in the tsconfig of repo-tools that I needed to carry over for this reason

**Review by bradzacher — DISMISSED** [review]
